stan/math/prim/core/complex_base.hpp

```diff
@@ -0,0 +1,275 @@
+#ifndef STAN_MATH_PRIM_CORE_COMPLEX_BASE_HPP
+#define STAN_MATH_PRIM_CORE_COMPLEX_BASE_HPP
+
+#include <stan/math/prim/fun/square.hpp>
+#include <complex>
+#include <type_traits>
+
+namespace stan {
+namespace math {
+
+/**
+ * Base class for complex numbers.  The template parameter `V` is the
+ * value type.  This class is used like the curiously recursive
+ * template pattern (CRTP) with derived type `complex<V>`.  Thus any
+ * extending class must be of the form `complex<V>`.
+ *
+ * @tparam V value type for extending complex class
+ */
+template <typename V>
+class complex_base {
+ protected:
+  /**
+   * Real part.
+   */
+  V re_;
+
+  /**
+   * Imaginary part.
+   */
+  V im_;
+
+ public:
+  /**
+   * Type of real and imaginary parts.
+   */
+  using value_type = V;
+
+  /**
+   * Derived complex type used for function return types.
+   */
+  using complex_type = std::complex<value_type>;
+
+  /**
+   * Construct a complex base with the specified real and imaginary
+   * parts.
+   *
+   * @tparam T real type (must be assignable to `V`)
+   * @tparam U imaginary type (must be assignable to `V`)
+   * @param[in] re real part
+   * @param[in] im imaginary part (default 0)
+   */
+  template <typename T, typename U>
+  complex_base(const T& re, const U& im) : re_(re), im_(im) {}
+
+  /**
+   * Construct a complex base with the specified real part and a zero
+   * imaginary part.
+   *
+   * @tparam T real type (must be assignable to `V`)
+   * @param[in] re real part
+   */
+  template <typename T>
+  complex_base(const T& re)  // NOLINT(runtime/explicit)
+      : complex_base(re, 0) {}
+
+  /**
+   * Construct a complex base with the specified real part and zero
+   * imaginary part.
+   */
+  complex_base() : complex_base(0, 0) {}
+
+  /**
+   * Constructs the complex number from the specified complex number
+   * of a different type.
+   *
+   * @param[in] other another complex to use as source
+   */
+  template <typename T>
+  complex_base(const std::complex<T>& other)  // NOLINT(runtime/explicit)
+      : complex_base(other.real(), other.imag()) {}
+
+  /**
+   * Destroy this complex number.
+   */
+  ~complex_base() {}
+
+  /**
+   * Return a reference to thi class cast to the derived complex
+   * type.
+   *
+   * @return reference to this class cast to the complex return type
+   */
+  complex_type& derived_complex_ref() {
+    return static_cast<complex_type&>(*this);
+  }
```

Comment:
I didn't do that because it's not an exact use of the CRTP.  It's just CRTP-like, so there's no template parameter `Derived`.  Instead there's a template parameter `T` and the equivalent of derived is `std::complex<T>`, which is typedef-ed as `complex_type`.  

But I'm happy to change it if you think that'd be clearer.

Reply:
I'm _okay_ with it as is, but imo I'd rather just have a CRTP than CRTP-like.